rasa/utils/tensorflow/data_generator.py

```diff
@@ -423,3 +426,66 @@ def _linearly_increasing_batch_size(self) -> int:
             )
         else:
             return int(self.batch_size[0])
+
+
+class MessageStreamDataGenerator(tf.keras.utils.Sequence):
+    """Data generator working with a message stream."""
+
+    def __init__(
+        self,
+        training_data: TrainingData,
+        batch_size: int,
+        message_to_data_function: Callable[[List[Message]], Data],
+        batch_strategy: Text = SEQUENCE,
+        shuffle: bool = True,
+    ):
+        """Initializes the data generator.
+
+        Args:
+            model_data: The model data to use.
+            batch_size: The batch size(s).
+            batch_strategy: The batch strategy.
+            shuffle: If 'True', data should be shuffled.
+        """
+        self.training_data = training_data
+        self.batch_size = batch_size
+        self.message_to_data_function = message_to_data_function
+        self.shuffle = shuffle
+        self.batch_strategy = batch_strategy
+        self.current_message_stream = None
+        self.on_epoch_end()
+
+    def __len__(self) -> int:
+        """Number of batches in the Sequence.
+
+        Returns:
+            The number of batches in the Sequence.
+        """
+        num_examples = len(self.training_data.training_examples)
+        batch_size = self.batch_size
+        return num_examples // batch_size + int(num_examples % batch_size > 0)
+
+    @lru_cache(1)  # quick fix, as index=0 is accessed twice somewhere
+    def __getitem__(self, index: int) -> Tuple[Any, Any]:
+        """Gets batch at position `index`.
+
+        Arguments:
+            index: position of the batch in the Sequence (ignored)
+
+        Returns:
+            A batch (tuple of input data and target data).
+        """
+        messages = []
+        while len(messages) < self.batch_size:
+            try:
+                messages.append(next(self.current_message_stream))
+            except StopIteration:
+                break
+        data = self.message_to_data_function(messages)
+        return RasaDataGenerator.prepare_batch(data), None
+
+    def on_epoch_end(self) -> None:
+        """Update the data after every epoch."""
+        self.current_message_stream = self.training_data.stream_featurized_messages(
```

Comment:
how would we e.g. shuffle chunks after epoch end?

Reply:
The TrainingData object could tell the featurereader to read the chunks in a random/different order. And then you still need to assure that the right message gets the right features. Some ID of the message that is also stored in the chunk or so.

Reply:
sweet, thanks for implementing this as well 💯